src/AnalyzeView/GeoTagController.cc

```diff
@@ -73,15 +76,191 @@ void GeoTagWorker::run(void)
+
+    ////////// Parse exif data
+    timerParseExif.start();
+
+    // Parse EXIF
+    ExifParser exifParser;
+    _tagTime.clear();
+    for (int i = 0; i < _imageBuffers.count(); i++) {
+         _tagTime.append(exifParser.readTime(_imageBuffers[i]));
+    }
+
+    qWarning() << "Parse exif data time elapsed: " << timerParseExif.elapsed() << " milliseconds";
+
+    ////////// Load PX4 log
+    timerLoadLogFile.start();
+
+    _geoRef.clear();
+    _triggerTime.clear();
+    if (!parsePX4Log()) {
+        qWarning() << "Geotagging failed";
+        return;
+    }
+    qWarning() << "Found " << _geoRef.count() << " trigger logs.";
+
+    qWarning() << "Log loading time elapsed: " << timerLoadLogFile.elapsed() << " milliseconds";
+
+    ////////// Filter Trigger
+    timerFilter.start();
+
+    if (!triggerFiltering()) {
+        qWarning() << "Geotagging failed";
+        return;
+    }
+
+    qWarning() << "Filter time elapsed: " << timerFilter.elapsed() << " milliseconds";
+
+    //////////// Tag images
+    timerGeotag.start();
+
+    for(int i = 0; i < _imageIndices.count() && i < _triggerIndices.count() && i < imageList.count(); i++) {
+        if (!exifParser.write(_imageBuffers[_imageIndices[i]], _geoRef[_triggerIndices[i]])) {
+            _cancel = true;
+            break;
+        } else {
+            QFile file(_imageDirectory + "/TAGGED/" + imageList[_imageIndices[i]].fileName());
+            if (file.open( QFile::WriteOnly)) {
+                file.write(_imageBuffers[_imageIndices[i]]);
+                file.close();
+            }
+        }
+    }
+
+    qWarning() << "Tagging images time elapsed: " << timerGeotag.elapsed() << " milliseconds";
+
     for (int i=0; i<10;i++) {
         if (_cancel) {
             emit error(tr("Tagging cancelled"));
             return;
         }
         emit progressChanged(i*10);
-        sleep(1);
+        //sleep(1);
     }
 
+    qWarning() << "Total time elapsed: " << timerTotal.elapsed() << " milliseconds";
     emit progressChanged(100);
     emit taggingComplete();
 }
+
+bool GeoTagWorker::parsePX4Log()
+{
+    // general message header
+    // char header[] = {0xA3, 0x95, 0x00};
+    // header for GPOS message
+    char gposHeader[] = {0xA3, 0x95, 0x10, 0x00};
+    int gposOffsets[3] = {3, 7, 11};
+    int gposLengths[3] = {4, 4, 4};
+    // header for trigger message
+    char triggerHeader[] = {0xA3, 0x95, 0x37, 0x00};
+    int triggerOffsets[2] = {3, 11};
+    int triggerLengths[2] = {8, 4};
+    // load log
+    QFile file(_logFile);
+    if (!file.open(QIODevice::ReadOnly)) {
+        qWarning() << "Could not open log file";
+        return false;
+    }
+    QByteArray log = file.readAll();
+    file.close();
+
+    // extract trigger data
+    int index = 1;
+    int sequence = -1;
+    QGeoCoordinate lastCoordinate;
+    while(index < log.count() - 1) {
+        int gposIndex = log.indexOf(gposHeader, index + 1);
```

Comment:
Not sure this mechanism will work. You could have random bytes in the log file which could match these header bytes, but they are just in the midlle of some other log message. Seems like you are going to need to parse through the log sequentially to avoid incorrect detection.